File: experiments/gilded-rose/ts/attempts/opus46high/4/gilded-rose.ts
export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name: string, sellIn: number, quality: number) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

interface ItemUpdater {
  update(item: Item): void;
}

class NormalUpdater implements ItemUpdater {
  update(item: Item): void {
    item.quality = Math.max(0, item.quality - (item.sellIn <= 0 ? 2 : 1));
    item.sellIn--;
  }
}

class AgedBrieUpdater implements ItemUpdater {
  update(item: Item): void {
    item.quality = Math.min(50, item.quality + (item.sellIn <= 0 ? 2 : 1));
    item.sellIn--;
  }
}

class BackstagePassUpdater implements ItemUpdater {
  update(item: Item): void {
    if (item.sellIn <= 0) {
      item.quality = 0;
    } else if (item.sellIn <= 5) {
      item.quality = Math.min(50, item.quality + 3);
    } else if (item.sellIn <= 10) {
      item.quality = Math.min(50, item.quality + 2);
    } else {
      item.quality = Math.min(50, item.quality + 1);
    }
    item.sellIn--;
  }
}

class SulfurasUpdater implements ItemUpdater {
  update(_item: Item): void { }
}

class ConjuredUpdater implements ItemUpdater {
  update(item: Item): void {
    item.quality = Math.max(0, item.quality - (item.sellIn <= 0 ? 4 : 2));
    item.sellIn--;
  }
}

function getUpdater(name: string): ItemUpdater {
  if (name === 'Aged Brie') return new AgedBrieUpdater();
  if (name === 'Backstage passes to a TAFKAL80ETC concert') return new BackstagePassUpdater();
  if (name === 'Sulfuras, Hand of Ragnaros') return new SulfurasUpdater();
  if (name.startsWith('Conjured')) return new ConjuredUpdater();
  return new NormalUpdater();
}

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  updateQuality() {
    for (const item of this.items) {
      getUpdater(item.name).update(item);
    }
    return this.items;
  }
}

File: experiments/gilded-rose/fsharp/attempts/opus46high/4/GildedRose.fs
module GildedRose

open System.Collections.Generic

type Item =
    { Name: string
      SellIn: int
      Quality: int }

let private clampQuality quality = max 0 (min 50 quality)

let private isConjured (name: string) = name.StartsWith("Conjured")

let private updateItem item =
    match item.Name with
    | "Sulfuras, Hand of Ragnaros" -> item
    | "Aged Brie" ->
        let quality =
            if item.SellIn <= 0 then
                item.Quality + 2
            else
                item.Quality + 1

        { item with
            SellIn = item.SellIn - 1
            Quality = clampQuality quality }
    | "Backstage passes to a TAFKAL80ETC concert" ->
        let quality =
            if item.SellIn <= 0 then
                0
            elif item.SellIn <= 5 then
                item.Quality + 3
            elif item.SellIn <= 10 then
                item.Quality + 2
            else
                item.Quality + 1

        { item with
            SellIn = item.SellIn - 1
            Quality = clampQuality quality }
    | name when isConjured name ->
        let degradation = if item.SellIn <= 0 then 4 else 2

        { item with
            SellIn = item.SellIn - 1
            Quality = clampQuality (item.Quality - degradation) }
    | _ ->
        let degradation = if item.SellIn <= 0 then 2 else 1

        { item with
            SellIn = item.SellIn - 1
            Quality = clampQuality (item.Quality - degradation) }

type GildedRoseShop(items: IList<Item>) =
    member _.Items = items

    member _.UpdateQuality() =
        for i = 0 to items.Count - 1 do
            items.[i] <- updateItem items.[i]
